### PR TITLE
[SPARK-44591][CONNECT][SQL] Add jobTags to SparkListenerSQLExecutionStart

### PR DESCRIPTION
**Pull request by jdesjean**
### What changes were proposed in this pull request?
Add jobTags to SparkListenerSQLExecutionStart

### Why are the changes needed?
As part [SPARK-43952](https://issues.apache.org/jira/browse/SPARK-43952), users can define job tags via SparkContext.addJobTag. These tags are then used to trigger cancelation via SparkContext.cancelJobsByTag. Furthermore, these tags can be used to logically group multiple jobs together.

Listener of job events can retrieve job tags via  SparkListenerJobStart.props.getProperty(SparkContext.SPARK_JOB_TAGS)

Listener of SQL events can link SparkListenerJobStart & SparkListenerSQLExecutionStart via SparkListenerJobStart.props.getProperty(SQLExecution.EXECUTION_ID_KEY).

However, some SQL executions do not trigger jobs (i.e. commands). As such listeners of SQL executions cannot resolve job tags of all executions.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
```
testOnly org.apache.spark.sql.execution.SQLExecutionSuite
```

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLExecutionSuite.scala [hunk 1]
```diff
@@ -203,6 +203,35 @@ class SQLExecutionSuite extends SparkFunSuite {
       spark.stop()
     }
   }
+
+  test("jobTags property.") {
```
Comment (juliuszsompolski) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLExecutionSuite.scala:207]: nit: remove "." from test name

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLExecutionSuite.scala [hunk 2]
```diff
@@ -203,6 +203,35 @@ class SQLExecutionSuite extends SparkFunSuite {
...
+    val spark = SparkSession.builder.master("local[*]").appName("test").getOrCreate()
+
+    // Disable Photon as the test uses a local spark session and hence JNI is not initialized.
```
Comment (juliuszsompolski) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLExecutionSuite.scala:210]: nit: this comment seems unrelated and accidentally copied over from somewhere.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLExecutionSuite.scala [hunk 3]
```diff
@@ -203,6 +203,35 @@ class SQLExecutionSuite extends SparkFunSuite {
...
+        }
+      }
+    })
+
+    spark.range(1).collect()
+
+    assert(jobTags.contains(jobTag))
+    assert(sqlJobTags.contains(jobTag))
+
+    spark.stop()
```
Comment (juliuszsompolski) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLExecutionSuite.scala:233]: other tests around here seem to wrap in try and do spark.stop() in finally.
Could you also do spark.sparkContext.removeJobTag(jobTag) in the finally, as otherwise I think the SparkContext gets shared between tests and the tag can get leaked into other tests?

**Review by juliuszsompolski — APPROVED** [review]
LGTM with some test nits.
cc @gengliangwang

**Review by juliuszsompolski — APPROVED** [review]
LGTM cc @gengliangwang

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala [hunk 4]
```diff
@@ -53,7 +53,8 @@ case class SparkListenerSQLExecutionStart(
     physicalPlanDescription: String,
     sparkPlanInfo: SparkPlanInfo,
     time: Long,
-    modifiedConfigs: Map[String, String] = Map.empty)
+    modifiedConfigs: Map[String, String] = Map.empty,
+    jobTags: Set[String] = Set.empty)
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala:57]: QQ: where will the jobTags in SparkListenerSQLExecutionStart event be used?

Comment (juliuszsompolski) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala:57]: As a followup to https://github.com/apache/spark/pull/41964, currently SQL Executions associated with a Spark Connect execution are found by using the EXECUTION_ID_KEY of corresponding Spark Jobs.
There are however various Spark Connect executions that start SQL Execution without starting a Spark Job. For example various commands like SHOW TABLES. Currently with https://github.com/apache/spark/pull/41964 the SQL Executions of these will not be linked from the Connect tab. With the jobTag, they can be now identified using the Spark Connect tag. Since this PR is almost ready to merge (hope CI will finish soon), I think using it for the Connect tab could be done in a followup PR?

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master/3.5
